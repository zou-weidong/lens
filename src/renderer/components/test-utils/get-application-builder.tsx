/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import type { IObservableArray, ObservableSet } from "mobx";
import { computed, observable, runInAction } from "mobx";
import React from "react";
import { Router } from "react-router";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import allowedResourcesInjectable from "../../cluster-frame-context/allowed-resources.injectable";
import type { RenderResult } from "@testing-library/react";
import { getByText, fireEvent } from "@testing-library/react";
import type { KubeResource } from "../../../common/rbac";
import type { DiContainer } from "@ogre-tools/injectable";
import clusterStoreInjectable from "../../../common/cluster-store/cluster-store.injectable";
import type { ClusterStore } from "../../../common/cluster-store/cluster-store";
import mainExtensionsInjectable from "../../../extensions/main-extensions.injectable";
import { pipeline } from "@ogre-tools/fp";
import { flatMap, compact, join, get, filter, map, matches, last } from "lodash/fp";
import preferenceNavigationItemsInjectable from "../+preferences/preferences-navigation/preference-navigation-items.injectable";
import navigateToPreferencesInjectable from "../../../common/front-end-routing/routes/preferences/navigate-to-preferences.injectable";
import type { MenuItemOpts } from "../../../main/menu/application-menu-items.injectable";
import applicationMenuItemsInjectable from "../../../main/menu/application-menu-items.injectable";
import type { MenuItemConstructorOptions, MenuItem } from "electron";
import storesAndApisCanBeCreatedInjectable from "../../stores-apis-can-be-created.injectable";
import navigateToHelmChartsInjectable from "../../../common/front-end-routing/routes/cluster/helm/charts/navigate-to-helm-charts.injectable";
import hostedClusterInjectable from "../../cluster-frame-context/hosted-cluster.injectable";
import { ClusterFrameContext } from "../../cluster-frame-context/cluster-frame-context";
import type { Cluster } from "../../../common/cluster/cluster";
import { KubeObjectStore } from "../../../common/k8s-api/kube-object.store";
import clusterFrameContextInjectable from "../../cluster-frame-context/cluster-frame-context.injectable";
import startMainApplicationInjectable from "../../../main/start-main-application/start-main-application.injectable";
import startFrameInjectable from "../../start-frame/start-frame.injectable";
import type { NamespaceStore } from "../+namespaces/store";
import namespaceStoreInjectable from "../+namespaces/store.injectable";
import historyInjectable from "../../navigation/history.injectable";
import type { MinimalTrayMenuItem } from "../../../main/tray/electron-tray/electron-tray.injectable";
import electronTrayInjectable from "../../../main/tray/electron-tray/electron-tray.injectable";
import applicationWindowInjectable from "../../../main/start-main-application/lens-window/application-window/application-window.injectable";
import { getDiForUnitTesting as getRendererDi } from "../../getDiForUnitTesting";
import { getDiForUnitTesting as getMainDi } from "../../../main/getDiForUnitTesting";
import { overrideChannels } from "../../../test-utils/channel-fakes/override-channels";
import trayIconPathsInjectable from "../../../main/tray/tray-icon-path.injectable";
import assert from "assert";
import { openMenu } from "react-select-event";
import userEvent from "@testing-library/user-event";
import lensProxyPortInjectable from "../../../main/lens-proxy/lens-proxy-port.injectable";
import type { Route } from "../../../common/front-end-routing/front-end-route-injection-token";
import type { NavigateToRouteOptions } from "../../../common/front-end-routing/navigate-to-route-injection-token";
import { navigateToRouteInjectionToken } from "../../../common/front-end-routing/navigate-to-route-injection-token";
import type { LensMainExtension } from "../../../extensions/lens-main-extension";
import type { LensExtension } from "../../../extensions/lens-extension";

import extensionInjectable from "../../../extensions/extension-loader/extension/extension.injectable";
import { renderFor } from "./renderFor";
import { RootFrame } from "../../frames/root-frame/root-frame";
import { ClusterFrame } from "../../frames/cluster-frame/cluster-frame";
import hostedClusterIdInjectable from "../../cluster-frame-context/hosted-cluster-id.injectable";

type Callback = (dis: DiContainers) => void | Promise<void>;

type EnableExtensions<T> = (...extensions: T[]) => void;
type DisableExtensions<T> = (...extensions: T[]) => void;

export interface ApplicationBuilder {
  dis: DiContainers;
  setEnvironmentToClusterFrame: () => ApplicationBuilder;

  extensions: {
    renderer: {
      enable: EnableExtensions<LensRendererExtension>;
      disable: DisableExtensions<LensRendererExtension>;
    };

    main: {
      enable: EnableExtensions<LensMainExtension>;
      disable: DisableExtensions<LensMainExtension>;
    };
  };

  allowKubeResource: (resourceName: KubeResource) => ApplicationBuilder;
  beforeApplicationStart: (callback: Callback) => ApplicationBuilder;
  beforeRender: (callback: Callback) => ApplicationBuilder;
  render: () => Promise<RenderResult>;

  tray: {
    click: (id: string) => Promise<void>;
    get: (id: string) => MinimalTrayMenuItem | null;
    getIconPath: () => string;
  };

  applicationMenu: {
    click: (path: string) => void;
  };

  preferences: {
    close: () => void;
    navigate: () => void;
    navigateTo: (route: Route<any>, params: Partial<NavigateToRouteOptions<any>>) => void;
    navigation: {
      click: (id: string) => void;
    };
  };

  helmCharts: {
    navigate: () => void;
  };

  select: {
    openMenu: (id: string) => void;
    selectOption: (menuId: string, labelText: string) => void;
  };
}

interface DiContainers {
  rendererDi: DiContainer;
  mainDi: DiContainer;
}

interface Environment {
  RootComponent: React.ElementType;
  onAllowKubeResource: () => void;
}

export const getApplicationBuilder = () => {
  const mainDi = getMainDi({
    doGeneralOverrides: true,
  });

  const overrideChannelsForWindow = overrideChannels(mainDi);

  const rendererDi = getRendererDi({
    doGeneralOverrides: true,
  });

  overrideChannelsForWindow(rendererDi);

  const dis = { rendererDi, mainDi };

  const clusterStoreStub = {
    provideInitialFromMain: () => {},
    getById: (): null => null,
  } as unknown as ClusterStore;

  rendererDi.override(clusterStoreInjectable, () => clusterStoreStub);
  rendererDi.override(storesAndApisCanBeCreatedInjectable, () => true);
  mainDi.override(clusterStoreInjectable, () => clusterStoreStub);

  const beforeApplicationStartCallbacks: Callback[] = [];
  const beforeRenderCallbacks: Callback[] = [];

  const rendererExtensionsState = observable.set<LensRendererExtension>();
  const mainExtensionsState = observable.set<LensMainExtension>();

  rendererDi.override(subscribeStoresInjectable, () => () => () => {});

  const environments = {
    application: {
      RootComponent: RootFrame,

      onAllowKubeResource: () => {
        throw new Error(
          "Tried to allow kube resource when environment is not cluster frame.",
        );
      },
    } as Environment,

    clusterFrame: {
      RootComponent: ClusterFrame,
      onAllowKubeResource: () => {},
    } as Environment,
  };

  let environment = environments.application;

  rendererDi.override(
    currentlyInClusterFrameInjectable,
    () => environment === environments.clusterFrame,
  );

  rendererDi.override(rendererExtensionsInjectable, () =>
    computed(() => [...rendererExtensionsState]),
  );

  mainDi.override(mainExtensionsInjectable, () =>
    computed(() => [...mainExtensionsState]),
  );

  let trayMenuIconPath: string;

  const traySetMenuItemsMock = jest.fn<any, [MinimalTrayMenuItem[]]>();

  mainDi.override(electronTrayInjectable, () => ({
    start: () => {
      const iconPaths = mainDi.inject(trayIconPathsInjectable);

      trayMenuIconPath = iconPaths.normal;
    },
    stop: () => {},
    setMenuItems: traySetMenuItemsMock,
    setIconPath: (path) => {
      trayMenuIconPath = path;
    },
  }));

  let allowedResourcesState: IObservableArray<KubeResource>;
  let rendered: RenderResult;

  const enableExtensionsFor = <T extends ObservableSet>(
    extensionState: T,
    di: DiContainer,
  ) => {
    const getExtension = (extension: LensExtension) =>
      di.inject(extensionInjectable, extension);

    return (...extensionInstances: LensExtension[]) => {
      const addAndEnableExtensions = () => {
        extensionInstances.forEach(instance => {
          const extension = getExtension(instance);

          extension.register();
        });

        runInAction(() => {
          extensionInstances.forEach((extension) => {
            extensionState.add(extension);
          });
        });
      };

      if (rendered) {
        addAndEnableExtensions();
      } else {
        builder.beforeRender(addAndEnableExtensions);
      }
    };
  };

  const builder: ApplicationBuilder = {
    dis,

    applicationMenu: {
      click: (path: string) => {
        const applicationMenuItems = mainDi.inject(
          applicationMenuItemsInjectable,
        );

        const menuItems = pipeline(
          applicationMenuItems.get(),
          flatMap(toFlatChildren(null)),
          filter((menuItem) => !!menuItem.click),
        );

        const menuItem = menuItems.find((menuItem) => menuItem.path === path);

        if (!menuItem) {
          const availableIds = menuItems.map(get("path")).join('", "');

          throw new Error(
            `Tried to click application menu item with ID "${path}" which does not exist. Available IDs are: "${availableIds}"`,
          );
        }

        menuItem.click?.(
          {
            menu: null as never,
            commandId: 0,
            userAccelerator: null,
            ...menuItem,
          } as MenuItem,
          undefined,
          {},
        );
      },
    },

    tray: {
      get: (id: string) => {
        const lastCall = last(traySetMenuItemsMock.mock.calls);

        assert(lastCall);

        return lastCall[0].find(matches({ id })) ?? null;
      },

      getIconPath: () => trayMenuIconPath,

      click: async (id: string) => {
        const lastCall = last(traySetMenuItemsMock.mock.calls);

        assert(lastCall);

        const trayMenuItems = lastCall[0];

        const menuItem = trayMenuItems.find(matches({ id })) ?? null;

        if (!menuItem) {
          const availableIds = pipeline(
            trayMenuItems,
            filter(item => !!item.click),
            map(item => item.id),
            join(", "),
          );

          throw new Error(`Tried to click tray menu item with ID ${id} which does not exist. Available IDs are: "${availableIds}"`);
        }

        if (!menuItem.enabled) {
          throw new Error(`Tried to click tray menu item with ID ${id} which is disabled.`);
        }

        await menuItem.click?.();
      },
    },

    preferences: {
      close: () => {
        const link = rendered.getByTestId("close-preferences");

        fireEvent.click(link);
      },

      navigate: () => {
        const navigateToPreferences = rendererDi.inject(navigateToPreferencesInjectable);

        navigateToPreferences();
      },

      navigateTo: (route: Route<any>, params: Partial<NavigateToRouteOptions<any>>) => {
        const navigateToRoute = rendererDi.inject(navigateToRouteInjectionToken);

        navigateToRoute(route, params);
      },

      navigation: {
        click: (id: string) => {
          const link = rendered.queryByTestId(`tab-link-for-${id}`);

          if (!link) {
            const preferencesNavigationItems = rendererDi.inject(
              preferenceNavigationItemsInjectable,
            );

            const availableIds = preferencesNavigationItems
              .get()
              .map(get("id"));

            throw new Error(
              `Tried to click navigation item "${id}" which does not exist in preferences. Available IDs are "${availableIds.join('", "')}"`,
            );
          }

          fireEvent.click(link);
        },
      },
    },

    helmCharts: {
      navigate: () => {
        const navigateToHelmCharts = rendererDi.inject(navigateToHelmChartsInjectable);

        navigateToHelmCharts();
      },
    },

    setEnvironmentToClusterFrame: () => {
      environment = environments.clusterFrame;

      allowedResourcesState = observable.array();

      rendererDi.override(allowedResourcesInjectable, () =>
        computed(() => new Set([...allowedResourcesState])),
      );

      const clusterStub = {
        accessibleNamespaces: [],
      } as unknown as Cluster;

      const namespaceStoreStub = {
        contextNamespaces: [],
        items: [],
      } as unknown as NamespaceStore;

      const clusterFrameContextFake = new ClusterFrameContext(
        clusterStub,

        {
          namespaceStore: namespaceStoreStub,
        },
      );

      rendererDi.override(namespaceStoreInjectable, () => namespaceStoreStub);
      rendererDi.override(hostedClusterInjectable, () => clusterStub);
      rendererDi.override(hostedClusterIdInjectable, () => "irrelevant-hosted-cluster-id");
      rendererDi.override(clusterFrameContextInjectable, () => clusterFrameContextFake);

      // Todo: get rid of global state.
      KubeObjectStore.defaultContext.set(clusterFrameContextFake);

      return builder;
    },

    extensions: {
      renderer: {
        enable: enableExtensionsFor(rendererExtensionsState, rendererDi),
        disable: disableExtensionsFor(rendererExtensionsState, rendererDi),
      },

      main: {
        enable: enableExtensionsFor(mainExtensionsState, mainDi),
        disable: disableExtensionsFor(mainExtensionsState, mainDi),
      },
    },

    allowKubeResource: (resourceName) => {
      environment.onAllowKubeResource();

      runInAction(() => {
        allowedResourcesState.push(resourceName);
      });

      return builder;
    },

    beforeApplicationStart(callback: (dis: DiContainers) => void) {
      beforeApplicationStartCallbacks.push(callback);

      return builder;
    },

    beforeRender(callback: (dis: DiContainers) => void) {
      beforeRenderCallbacks.push(callback);

      return builder;
    },

    async render() {
      mainDi.inject(lensProxyPortInjectable).set(42);

      for (const callback of beforeApplicationStartCallbacks) {
        await callback(dis);
      }

      const startMainApplication = mainDi.inject(startMainApplicationInjectable);

      await startMainApplication();

      const applicationWindow = mainDi.inject(applicationWindowInjectable);

      await applicationWindow.start();

      const startFrame = rendererDi.inject(startFrameInjectable);

      await startFrame();

      for (const callback of beforeRenderCallbacks) {
        await callback(dis);
      }

      const history = rendererDi.inject(historyInjectable);

      const render = renderFor(rendererDi);

      rendered = render(
        <Router history={history}>
          <environment.RootComponent />
        </Router>,
      );

      return rendered;
    },

    select: {
      openMenu: (menuId) => {
        const selector = rendered.container.querySelector<HTMLElement>(
          `#${menuId}`,
        );

        assert(selector);

        openMenu(selector);
      },

      selectOption: (menuId, labelText) => {
        const menuOptions = rendered.baseElement.querySelector<HTMLElement>(
          `.${menuId}-options`,
        );

        assert(menuOptions);

        const option = getByText(menuOptions, labelText);

        userEvent.click(option);
      },
    },
  };

  return builder;
};

export type ToFlatChildren = (opts: MenuItemConstructorOptions) => (MenuItemOpts & { path: string })[];

function toFlatChildren(parentId: string | null | undefined): ToFlatChildren {
  return ({ submenu = [], ...menuItem }) => [
    {
      ...menuItem,
      path: pipeline([parentId, menuItem.id], compact, join(".")),
    },
    ...(
      Array.isArray(submenu)
        ? submenu.flatMap(toFlatChildren(menuItem.id))
        : [{
          ...submenu,
          path: pipeline([parentId, menuItem.id], compact, join(".")),
        }]
    ),
  ];
}

const disableExtensionsFor = <T extends ObservableSet>(
  extensionState: T,
  di: DiContainer,
) => {
  const getExtension = (instance: LensExtension) =>
    di.inject(extensionInjectable, instance);

  return (...extensionInstances: LensExtension[]) => {
    extensionInstances.forEach((instance) => {
      const extension = getExtension(instance);

      runInAction(() => {
        extension.deregister();

        extensionState.delete(instance);
      });
    });
  };
};
