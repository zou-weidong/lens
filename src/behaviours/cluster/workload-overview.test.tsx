/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { RenderResult } from "@testing-library/react";
import navigateToWorkloadsOverviewInjectable from "../../common/front-end-routing/routes/cluster/workloads/overview/navigate-to-workloads-overview.injectable";
import { type ApplicationBuilder, getApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";

describe("workload overview", () => {
  let rendered: RenderResult;
  let applicationBuilder: ApplicationBuilder;

  beforeEach(async () => {
    applicationBuilder = getApplicationBuilder().setEnvironmentToClusterFrame();
    applicationBuilder.allowKubeResource("pods");
    rendered = await applicationBuilder.render();
  });

  describe("when navigating to workload overview", () => {
    beforeEach(() => {
      applicationBuilder.navigateWith(navigateToWorkloadsOverviewInjectable);
    });

    it("renders", () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    it("does not yet show workload overview", () => {
      expect(rendered.queryByTestId("page-for-workloads-overview")).toBeInTheDocument();
    });

    it("shows pods pie chart", () => {
      expect(rendered.queryByTestId("workload-overview-status-chart-pods")).toBeInTheDocument();
    });
  });
});
