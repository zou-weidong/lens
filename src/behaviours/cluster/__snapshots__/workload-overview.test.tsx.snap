// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workload overview when navigating to workload overview renders 1`] = `
<body>
  <div>
    <div
      class="flex flex-col"
      data-testid="cluster-sidebar"
    >
      <div
        class="SidebarCluster"
      >
        <div
          class="Avatar rounded loadingAvatar"
          style="width: 40px; height: 40px;"
        >
          ??
        </div>
        <div
          class="loadingClusterName"
        />
      </div>
      <div
        class="sidebarNav sidebar-active-status"
      >
        <div
          class="SidebarItem"
          data-is-active-test="true"
          data-testid="sidebar-item-workloads"
        >
          <a
            aria-current="page"
            class="nav-item flex gaps align-center expandable active"
            data-testid="sidebar-item-link-for-workloads"
            href="/"
          >
            <i
              class="Icon svg focusable"
            >
              <span
                class="icon"
              />
            </i>
            <span
              class="link-text box grow"
            >
              Workloads
            </span>
            <i
              class="Icon expand-icon box right material focusable"
            >
              <span
                class="icon"
                data-icon-name="keyboard_arrow_down"
              >
                keyboard_arrow_down
              </span>
            </i>
          </a>
        </div>
        <div
          class="SidebarItem"
          data-is-active-test="false"
          data-testid="sidebar-item-config"
        >
          <a
            class="nav-item flex gaps align-center"
            data-testid="sidebar-item-link-for-config"
            href="/"
          >
            <i
              class="Icon material focusable"
            >
              <span
                class="icon"
                data-icon-name="list"
              >
                list
              </span>
            </i>
            <span
              class="link-text box grow"
            >
              Config
            </span>
          </a>
        </div>
        <div
          class="SidebarItem"
          data-is-active-test="false"
          data-testid="sidebar-item-network"
        >
          <a
            class="nav-item flex gaps align-center expandable"
            data-testid="sidebar-item-link-for-network"
            href="/"
          >
            <i
              class="Icon material focusable"
            >
              <span
                class="icon"
                data-icon-name="device_hub"
              >
                device_hub
              </span>
            </i>
            <span
              class="link-text box grow"
            >
              Network
            </span>
            <i
              class="Icon expand-icon box right material focusable"
            >
              <span
                class="icon"
                data-icon-name="keyboard_arrow_down"
              >
                keyboard_arrow_down
              </span>
            </i>
          </a>
        </div>
        <div
          class="SidebarItem"
          data-is-active-test="false"
          data-testid="sidebar-item-storage"
        >
          <a
            class="nav-item flex gaps align-center"
            data-testid="sidebar-item-link-for-storage"
            href="/"
          >
            <i
              class="Icon material focusable"
            >
              <span
                class="icon"
                data-icon-name="storage"
              >
                storage
              </span>
            </i>
            <span
              class="link-text box grow"
            >
              Storage
            </span>
          </a>
        </div>
        <div
          class="SidebarItem"
          data-is-active-test="false"
          data-testid="sidebar-item-helm"
        >
          <a
            class="nav-item flex gaps align-center expandable"
            data-testid="sidebar-item-link-for-helm"
            href="/"
          >
            <i
              class="Icon svg focusable"
            >
              <span
                class="icon"
              />
            </i>
            <span
              class="link-text box grow"
            >
              Helm
            </span>
            <i
              class="Icon expand-icon box right material focusable"
            >
              <span
                class="icon"
                data-icon-name="keyboard_arrow_down"
              >
                keyboard_arrow_down
              </span>
            </i>
          </a>
        </div>
        <div
          class="SidebarItem"
          data-is-active-test="false"
          data-testid="sidebar-item-user-management"
        >
          <a
            class="nav-item flex gaps align-center"
            data-testid="sidebar-item-link-for-user-management"
            href="/"
          >
            <i
              class="Icon material focusable"
            >
              <span
                class="icon"
                data-icon-name="security"
              >
                security
              </span>
            </i>
            <span
              class="link-text box grow"
            >
              Access Control
            </span>
          </a>
        </div>
        <div
          class="SidebarItem"
          data-is-active-test="false"
          data-testid="sidebar-item-custom-resources"
        >
          <a
            class="nav-item flex gaps align-center expandable"
            data-testid="sidebar-item-link-for-custom-resources"
            href="/"
          >
            <i
              class="Icon material focusable"
            >
              <span
                class="icon"
                data-icon-name="extension"
              >
                extension
              </span>
            </i>
            <span
              class="link-text box grow"
            >
              Custom Resources
            </span>
            <i
              class="Icon expand-icon box right material focusable"
            >
              <span
                class="icon"
                data-icon-name="keyboard_arrow_down"
              >
                keyboard_arrow_down
              </span>
            </i>
          </a>
        </div>
      </div>
    </div>
    <div
      class="TabLayout"
      data-testid="tab-layout"
    >
      <div
        class="Tabs center scrollable"
      >
        <div
          class="Tab flex gaps align-center active"
          data-is-active-test="true"
          data-testid="tab-link-for-overview"
          role="tab"
          tabindex="0"
        >
          <div
            class="label"
          >
            Overview
          </div>
        </div>
        <div
          class="Tab flex gaps align-center"
          data-is-active-test="false"
          data-testid="tab-link-for-pods"
          role="tab"
          tabindex="0"
        >
          <div
            class="label"
          >
            Pods
          </div>
        </div>
      </div>
      <main>
        <div
          class="WorkloadsOverview flex column gaps"
          data-testid="page-for-workloads-overview"
        >
          <div
            class="header flex gaps align-center"
          >
            <h5
              class="box grow"
            >
              Overview
            </h5>
            <div>
              <div
                class="Select theme-dark NamespaceSelect NamespaceSelectFilter css-b62m3t-container"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-overview-namespace-select-filter-input-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                />
                <div
                  class="Select__control css-1s2u09g-control"
                >
                  <div
                    class="Select__value-container Select__value-container--is-multi css-319lph-ValueContainer"
                  >
                    <div
                      class="Select__placeholder css-14el2xx-placeholder"
                      id="react-select-overview-namespace-select-filter-input-placeholder"
                    >
                      All namespaces
                    </div>
                    <div
                      class="Select__input-container css-6j8wv5-Input"
                      data-value=""
                    >
                      <input
                        aria-autocomplete="list"
                        aria-describedby="react-select-overview-namespace-select-filter-input-placeholder"
                        aria-expanded="false"
                        aria-haspopup="true"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        class="Select__input"
                        id="overview-namespace-select-filter-input"
                        role="combobox"
                        spellcheck="false"
                        style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="Select__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="Select__indicator-separator css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="Select__indicator Select__dropdown-indicator css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="OverviewStatuses"
          >
            <div
              class="workloads"
            >
              <div
                class="workload"
              >
                <div
                  class="title"
                >
                  <a>
                    Pods (0)
                  </a>
                </div>
                <div
                  class="OverviewWorkloadStatus"
                >
                  <div
                    class="flex column align-center box grow"
                  >
                    <div
                      class="Chart PieChart flex column align-center"
                      data-testid="workload-overview-status-chart-pods"
                    >
                      <div
                        class="chart-container"
                      >
                        <canvas
                          class="chartjs-render-monitor"
                          height="0"
                          style="display: block;"
                          width="0"
                        />
                        <div
                          class="chartjs-tooltip flex column"
                        />
                      </div>
                      <div
                        class="legend flex wrap"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </main>
    </div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
</body>
`;
